File: kth smallest elementin bst.py
#TC: O(n) in worst case if the element is last element


# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def kthSmallest(self, root, k):
        """
        :type root: TreeNode
        :type k: int
        :rtype: int
        """
        self.count=0
        self.result=0
        def inorder(root,k):
            #base
            if root==None: return
            #logic
            inorder(root.left,k)
            self.count+=1
            if self.count==k: 
                self.result=root.val
                return
            inorder(root.right,k)
         
        inorder(root,k)
        return self.result
        
#using int 
'''
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
import sys
class Solution(object):
    def kthSmallest(self, root, k):
        """
        :type root: TreeNode
        :type k: int
        :rtype: int
        """
        self.count=k
        self.result=0
        def inorder(root,k):
            #base
            if root==None: return sys.maxsize
            #logic
            left=inorder(root.left,k)
            if (left!=sys.maxsize): return left
            self.count-=1
            
            if self.count==0: 
                return root.val
            return inorder(root.right,k)
         
        return inorder(root,k)
'''

#This code is written in Java for dynamic tree and watch the video.
'''
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
  public class Solution {
        public int kthSmallest(TreeNode root, int k) {
            TreeNodeWithCount rootWithCount = buildTreeWithCount(root);
            return kthSmallest(rootWithCount, k);
        }
        
        private TreeNodeWithCount buildTreeWithCount(TreeNode root) {
            if (root == null) return null;
            TreeNodeWithCount rootWithCount = new TreeNodeWithCount(root.val);
            rootWithCount.left = buildTreeWithCount(root.left);
            rootWithCount.right = buildTreeWithCount(root.right);
            if (rootWithCount.left != null) rootWithCount.count += rootWithCount.left.count;
            if (rootWithCount.right != null) rootWithCount.count += rootWithCount.right.count;
            return rootWithCount;
        }
        
        private int kthSmallest(TreeNodeWithCount rootWithCount, int k) {
            if (k <= 0 || k > rootWithCount.count) return -1;
            if (rootWithCount.left != null) {
                if (rootWithCount.left.count >= k) return kthSmallest(rootWithCount.left, k);
                if (rootWithCount.left.count == k-1) return rootWithCount.val;
                return kthSmallest(rootWithCount.right, k-1-rootWithCount.left.count);
            } else {
                if (k == 1) return rootWithCount.val;
                return kthSmallest(rootWithCount.right, k-1);
            }
        }
        
        class TreeNodeWithCount {
            int val;
            int count;
            TreeNodeWithCount left;
            TreeNodeWithCount right;
            TreeNodeWithCount(int x) {val = x; count = 1;};
        }
    }
'''

File: src/com/madhurima/KthSmallestElementInBST.java
// Time Complexity : O(n) worst case will visit all nodes
// Space Complexity : O(height of tree) (in case of recursion, its the recursion stack and in case of stack, its height of stack)
// Did this code successfully run on Leetcode : yes

package com.madhurima;

import java.util.Stack;

public class KthSmallestElementInBST {

}

 class TreeNode {
      int val;
      TreeNode left;
      TreeNode right;
      TreeNode() {}
      TreeNode(int val) { this.val = val; }
      TreeNode(int val, TreeNode left, TreeNode right) {
          this.val = val;
          this.left = left;
          this.right = right;
      }
  }

class SolutionRecursion {

    int count;
    int result;

    //recursive in-order travelsal
    public int kthSmallest(TreeNode root, int k) {
        if(root == null) return 0;

        result = -1;
        dfs(root, k);
        return result;

    }

    private void dfs(TreeNode root, int k){
        //base
        if(root ==  null){
            return;
        }

        //logic
        if(result == -1){ // if you already have the result, don't continue with the recursion
            dfs(root.left, k);
        }


        count++;
        if(count == k){
            result = root.val;
            return;
        }

        if(result == -1){
            dfs(root.right, k);
        }
    }
}

class SolutionIterative {

    //iterative approach
    public int kthSmallest(TreeNode root, int k) {
        if(root == null){
            return 0;
        }

        Stack<TreeNode> s = new Stack<>();
        int count = 0;

        while(root != null || !s.isEmpty()){
            while(root != null){
                s.push(root);
                root = root.left;
            }

            root = s.pop();
            count++;

            if(count == k){
                return root.val;
            }

            root = root.right;

        }

        return count;
    }
}
